### PR TITLE
Implement cellCharter

### LINKED ISSUE #7: method_CellCharter
**Update metadata:**
https://docs.google.com/spreadsheets/d/1QCeAF4yQG4bhZSGPQwwVBj_XF7ADY_2mK5xivAIfHsc/edit#gid=0

**Details of method (publication, GitHub):**
https://www.biorxiv.org/content/10.1101/2023.01.10.523386v1
https://github.com/CSOgroup/cellcharter

### PR DESCRIPTION
**Pull request by Jieran-S**
Cellcharter is implemented and tested. Should be ready to merge #7

### CODE REVIEW

File: method/CellCharter/CellCharter.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+    return adata
+
+## Your code goes here
+import cellcharter as cc
+import anndata as ad
+import scanpy as sc
+import pandas as pd
+import numpy as np
+from squidpy._constants._pkg_constants import Key
```
Comment (niklasmueboe) [inline method/CellCharter/CellCharter.py:125]: importing from private modules seems dangerous. Also I can't find this used anywhere?

Probably tis import should be removed

File: method/CellCharter/CellCharter.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+Dst = csr_matrix((dists, Adj.indices, Adj.indptr), shape=(N, N))
+Dst.setdiag(0.0)
+
+# Fitting into the respective data slot
+adata.obsp["spatial_distances"] = Dst
+adata.uns["spatial_neighbors"] = {
+        "connectivities_key": "spatial_connectivities",
+        "distances_key": "spatial_distances",
+        "params": {"coord_type": "generic", "delaunay": True, "transform": None},
```
Comment (niklasmueboe) [inline method/CellCharter/CellCharter.py:158]: These params might not always be correct. Is this used in later steps of the algorithm?

**Review by niklasmueboe — CHANGES_REQUESTED** [review]

File: method/CellCharter/CellCharter.py (hunk 1 above)
Comment (Jieran-S) [reply method/CellCharter/CellCharter.py:125]: Hi yes sorry this import will be removed!

File: method/CellCharter/CellCharter.py (hunk 2 above)
Comment (Jieran-S) [reply method/CellCharter/CellCharter.py:158]: Hi, these parameters aren't really used in any of the downstream analyses. I simply put in there as the code for calculating the spatial distance is taken from the case where `delaunay=True` and `coord_type=generic` in [`squidpy.gr.spatial_neighbors`](https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_neighbors.html). 

The main point of code lines 143-159 is to replicate how the squidpy function generates the distance matrix from the adjacency matrix. And here is what each parameter is for: 
* `delaunay=True`: It is set to be `True` because in cellcharter tutorial they mentioned the delaunay parameters need to be true for downstream functions. I also adapted the code for the distance matrix assuming `delaunay=True` in the function.
* `coord_type=generic`: If `delaunay=True`, this parameter doesn't make a difference, set to default
* `transform=None`: Set to default, assuming no transformation on the adjacency matrix.

Hope it answers the question! Those parameters don't affect the subsequent analysis but the dictionary structure needs to be there. Let me know how you think we could change it! :))

Comment (niklasmueboe) [reply method/CellCharter/CellCharter.py:157]: So do they always assume delaunay triangulation as input for the neighbor structure? Because they also do this n-th percentile trimming of the neighbor graph which does not necessarily make sense for some other methods, e.g. anything that is based of neighbors on a regular grid which is usually used for Visium for example.

Comment (Jieran-S) [reply method/CellCharter/CellCharter.py:157]: Ohh that's a good point! Sorry I missed that earlier. In their documentation, they only showcase the example of CosMx and they use `delaunay=True`. But in their [paper](https://www.nature.com/articles/s41588-023-01588-4#Sec7) they did mention that removing long links is unnecessary for regular structure technology. 
I will make it optional based on the `technology` option. If `technology==Visium`, I will skip the distance calculation and the `remove_long_links` part, as the distance matrix is only needed for`remove_long_links`.

**Comment by Jieran-S** [comment]
Another thing is, cellcharter uses scVI for dimensional reduction and batch correction, it was mentioned in their paper to be an important part. I did not implement that part as we will already provide a `dim_red` slot. Should I maybe make it an option for the scVI batch correction and use config file to activate it, or should I assume our preprocessing will test out all possibilities?

**Comment by niklasmueboe** [comment]
> Another thing is, cellcharter uses scVI for dimensional reduction and batch correction, it was mentioned in their paper to be an important part. I did not implement that part as we will already provide a `dim_red` slot. Should I maybe make it an option for the scVI batch correction and use config file to activate it, or should I assume our preprocessing will test out all possibilities?

bacth correction should nly be necessary for multi-sample analysis which we haven't implemented for now. And regarding dim-red I think we can always add scVI as additional preprocessing step alternatively to PCA, ...

So I think we can leave it as is for now

File: method/CellCharter/CellCharter.py (hunk 2 above)
Comment (niklasmueboe) [reply method/CellCharter/CellCharter.py]: It should also be done for "ST" not only "Visium".

The issue I see that it is technically not only dependent on the technology but also on the neighbor definition. e.g. you could define neighbors based on distance and then doing a trimming on the 99th percentile does not make much sense.

Not quite sure what the best approach is for now.

Comment (Jieran-S) [reply method/CellCharter/CellCharter.py]: Since the tirmming only makes sense when delaunay triangulation is applied to construct the neighborhood graph (hence not depending on the distance), I can try to implement some codes to check if the neighbor graphs provided fulfills delaunay criteria and proceed on trimming based on that. Do you think that'd be a way to work around it? :))

**Comment by Jieran-S** [comment]
Hi, I just talked to the author of cellcharter and updated our implementation: 
1. Added cellcharter support for raw data input. This is important as the dimensional reduction embedding provided by scVI is essential to cellcharter performance. 
2. Simply restrict the "remove_long_link" function to technology types(i.e. not applied to Visium and ST). The author mentioned that removing the top 1% or not actually do not affect the performance much. Even if the graph was not generated via delaunay triangulation, removing the top 1% wouldnt have too much of an impact.

**Comment by Jieran-S** [comment]
Change CC to accepting raw count input only. So:

1. Delaunay neighbor graph cut issue resolved
2. scVI is important to cellcharter. Can't be replaced by other dim-red preprocessing.

File: method/CellCharter/CellCharter.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+# Set up scvi model for reduced_dimension embedding
+scvi.settings.seed = seed
+scvi.model.SCVI.setup_anndata(
+    adata,
+    layer="counts"
+)
+
+model = scvi.model.SCVI(adata)
+model.train(early_stopping=True, enable_progress_bar=True)
```
Comment (niklasmueboe) [inline method/CellCharter/CellCharter.py:152]: Is this an interactive progress bar? Then it should probably be set to false?

File: method/CellCharter/CellCharter.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                          sample_key=None)
+
+# clustering
+model = cc.tl.Cluster(
+    n_clusters=n_clusters, 
+    random_state=seed,
+    convergence_tolerance=config["convergence_tolerance"], 
+    covariance_regularization=config["covariance_regularization"]
+    # if running on gpu
+    #trainer_params=dict(accelerator='gpu', devices=1)
```
Comment (niklasmueboe) [inline method/CellCharter/CellCharter.py:182]: So can we currently use GPU for this tool?

File: method/CellCharter/CellCharter.py (hunk 3 above)
Comment (Jieran-S) [reply method/CellCharter/CellCharter.py]: Tune them off! :))

File: method/CellCharter/CellCharter.py (hunk 4 above)
Comment (Jieran-S) [reply method/CellCharter/CellCharter.py]: Yes -- I added a cuda of GPU availability now.

**Review by niklasmueboe — APPROVED** [review]
